pallets/parachain-staking/src/tests.rs

```diff
@@ -4528,6 +4500,307 @@ fn delegation_events_convey_correct_position() {
+
+			assert!(! <DelayedPayouts<Test>>::contains_key(2));
+			assert!(! <Points<Test>>::contains_key(2), "We never rewarded points for round 2");
+			assert!(<Staked<Test>>::contains_key(2));
+
+			assert!(! <DelayedPayouts<Test>>::contains_key(3));
+			assert!(! <Points<Test>>::contains_key(3), "We never awarded points for round 3");
+			assert!(<Staked<Test>>::contains_key(3));
+
+			// collator 1 has been paid in this last block and associated storage cleaned up
+			assert!(! <AtStake<Test>>::contains_key(1, 1));
+			assert!(! <AwardedPts<Test>>::contains_key(1, 1));
+
+			// but collator 2 hasn't been paid
+			assert!(<AtStake<Test>>::contains_key(1, 2));
+			assert!(<AwardedPts<Test>>::contains_key(1, 2));
+
+			round = 4;
+			roll_to_round_begin(round.into());
+			expected.append(&mut vec![
+				Event::Rewarded(2, 1), // from previous round
+				Event::CollatorChosen(round, 1, 20),
+				Event::CollatorChosen(round, 2, 20),
+				Event::NewRound(15, round, 2, 40),
+			]);
+			assert_eq_events!(expected);
+
+			roll_to_round_end(4);
+			// collators have both been paid and storage fully cleaned up for round 1
+			assert!(! <AtStake<Test>>::contains_key(1, 2));
+			assert!(! <AwardedPts<Test>>::contains_key(1, 2));
+			assert!(! <Staked<Test>>::contains_key(1));
+			assert!(! <Points<Test>>::contains_key(1)); // points should be cleaned up
+			assert!(! <DelayedPayouts<Test>>::contains_key(1));
+
+			// no more events expected
+			assert_eq_events!(expected);
+
+		});
+}
+
+#[test]
+fn deferred_payment_steady_state_event_flow() {
+	use frame_support::traits::{
+		ExistenceRequirement,
+		WithdrawReasons,
+		Currency,
+		Imbalance,
+	};
+
+	// this test "flows" through a number of rounds, asserting that certain things do/don't happen
+	// once the staking pallet is in a "steady state" (specifically, once we are past the first few
+	// rounds to clear RewardPaymentDelay)
+
+	ExtBuilder::default()
+		.with_balances(vec![
+			   (  1, 200), ( 2, 200), ( 3, 200), ( 4, 200), // collators
+			   ( 11, 200), (22, 200), (33, 200), (44, 200), // delegators
+			   (111, 1000), // burn account, see `reset_issuance()`
+		])
+		.with_candidates(vec![(1, 200), (2, 200), (3, 200), (4, 200)])
+		.with_delegations(vec![
+			  (11, 1, 100), (11, 2, 100), // delegator 11 delegates 100 to 1 and 2
+			  (22, 2, 100), (22, 3, 100), // delegator 22 delegates 100 to 2 and 3
+			  (33, 3, 100), (33, 4, 100), // delegator 33 delegates 100 to 3 and 4
+			  (44, 4, 100), (44, 1, 100), // delegator 44 delegates 100 to 4 and 1
+		])
+		.build()
+		.execute_with(|| {
+			// convenience to set the round points consistently
+			let set_round_points = |round: u64| {
+				set_author(round as u32, 1, 1);
+				set_author(round as u32, 2, 1);
+				set_author(round as u32, 3, 1);
+				set_author(round as u32, 4, 1);
+			};
+
+			// grab initial issuance -- we will reset it before round issuance is calculated so that
+			// it is consistent every round
```

Comment:
in practice the round issuance will not be consistent every round? if so, why do we test it like this? It feels like we're just doing

```rust
for _ in 1..10 {
   unit_test()
}
```
what is the point of increasing the round number if we are resetting issuance after every round?

Reply:
The point of this test was to show that rounds are consistently following the same pattern. The test didn't turn out quite the way I wanted it to, but I think it's useful. It definitely helped find some minor issues in the pallet.

You're right that issuance won't be consistent every round -- this was quite a hack that I wasn't anticipating...

Reply:
> The point of this test was to show that rounds are consistently following the same pattern. 

If you reset issuance after every round and do the same changes + same checks, it's just the same test run multiple times.

Nothing is changing in every loop except the round number so the test is just verifying that the round number changing doesn't effect payouts?

Reply:
It ensures that the rounds are consistent once we reach `RewardPaymentDelay`. This could be done without looping, but would be harder to understand (arguably, I guess). I agree it's a little unorthodox...